File: Chapter 02/AbstractFactoryPatternExample.java
interface Dog {
    void draw();
}
class Corgi implements Dog {
    public void draw() {
        System.out.println("Drawing a Corgi.");
    }
}
class Retriever implements Dog {
    public void draw() {
        System.out.println("Drawing a Retriever.");
    }
}


interface Color {
    void fill();
}
class Gold implements Color {
    public void fill() {
        System.out.println("Filling with Gold color.");
    }
}
class Tricolor implements Color {
    public void fill() {
        System.out.println("Filling with Tricolor color.");
    }
}


interface AbstractFactory {
    Dog createDog();
    Color createColor();
}
class DogFactory implements AbstractFactory {
    public Dog createDog() {
        return new Corgi();
    }
    public Color createColor() {
        return new TriColor();
    }
}
class ColorFactory implements AbstractFactory {
    public Dog createDog() {
        return new Retriever();
    }
    public Color createColor() {
        return new Gold();
    }
}


public class AbstractFactoryPatternExample {
    public static void main(String[] args) {
        AbstractFactory dogFactory = new DogFactory();
        Dog corgi = dogFactory.createDog();
        Color tricolor = dogFactory.createColor();

        corgi.draw();
        tricolor.fill();

        AbstractFactory colorFactory = new ColorFactory();
        Dog retriever = colorFactory.createDog();
        Color gold = colorFactory.createColor();

        retriever.draw();
        gold.fill();
    }
}

File: Chapter 02/AbstractFactoryPatternExample.py
class Dog:
    def draw(self):
        pass
class Corgi(Dog):
    def draw(self):
        print("Drawing a Corgi.")
class Retriever(Dog):
    def draw(self):
        print("Drawing a Retriver.")

class Color:
    def fill(self):
        pass
class Tricolor(Color):
    def fill(self):
        print("Filling with Tricolor color.")
class Gold(Color):
    def fill(self):
        print("Filling with Gold color.")

class AbstractFactory:
    def create_dog(self):
        pass
    def create_color(self):
        pass

class DogFactory(AbstractFactory):
    def create_dog(self):
        return Corgi()
    def create_color(self):
        return Tricolor()

class ColorFactory(AbstractFactory):
    def create_dog(self):
        return Retriever()
    def create_color(self):
        return Gold()

if __name__ == "__main__":
    dog_factory = DogFactory()
    corgi = dog_factory.create_dog()
    tricolor = dog_factory.create_color()
    corgi.draw()
    tricolor.fill()

    color_factory = ColorFactory()
    retriever = color_factory.create_dog()
    gold = color_factory.create_color()
    retriever.draw()
    gold.fill()
